tensorboard/dataclass_compat.py

```diff
@@ -84,46 +91,61 @@ def _migrate_summary_event(event):
     return (event,)
 
 
-def _migrate_value(value):
+def _migrate_value(value, initial_metadata):
     """Convert an old value to a stream of new values. May mutate."""
-    if value.metadata.data_class != summary_pb2.DATA_CLASS_UNKNOWN:
+    metadata = initial_metadata.get(value.tag)
+    initial = False
+    if metadata is None:
+        initial = True
+        # Retain a copy of the initial metadata, so that even after we
+        # update its data class we know whether to also transform later
+        # events in this time series.
+        metadata = summary_pb2.SummaryMetadata()
+        metadata.CopyFrom(value.metadata)
+        initial_metadata[value.tag] = metadata
+    if metadata.data_class != summary_pb2.DATA_CLASS_UNKNOWN:
```

Comment:
Just to check my understanding, our assumption is that any summary that natively sets the data class on the initial metadata will already be using the dataclass-aware format (so to later summaries with that tag will require value conversions)?

Reply:
Yeah. In some sense there’s a bit of awkward coupling: one could imagine
factoring the migrations more finely as

 1. `data_compat` (v1 → v2 summaries)
 2. audio summary projection (label removal)
 3. hparams tensor fix
 4. `dataclass_compat` (add `metadata.data_class` only)

where there are some dependencies (4 depends on 2, 2 depends on 1) but
the dependency graph is not (an orientation of) a complete graph. Then
if we needed to add more read-time migrations later even after summaries
natively write `data_class` we’d still be able to do so.

For now, at least, I thought that it was simpler to keep everything in
this layer so that there are fewer moving pieces.